kstatus/README.md

```diff
@@ -0,0 +1,116 @@
+# kstatus
+
+kstatus provides tools for checking the status of Kubernetes resources. The primary use case is around the pull-apply loop
+where a set of resources are updated/created with `apply` and we want to know when the changes has either fully
+been reconciled in a cluster or something is causing the process to stall.
+
+## Concepts
+
+This effort has several goals, some with a shorter timeline than others. Initially, we want to provide
+a library that makes it easier to decide when changes to a set of resources have been reconciled in a cluster.
+Since the built-in kubernetes types doesn't provide this information directly, we will initially rely on rules
+specific for each type. For example, the library already contains rules for many of the most common built-in types 
+such a Deployment and StatefulSet.
+
+For custom types (CRDs), there currently isn't much guidance on which properties should be exposed in the status
+object and which conditions should be used. As part of this effort we want to define a set of standard conditions
+that the library will understand and that we encourage developers to adopt in their CRDs. These standard conditions will
+be focused on providing the necessary information for understanding status of the reconcile after `apply` and it is not
+expected that these will necessarily be the only conditions exposed by a CRD. Developers will be free to add as many conditions
+as they wish, but if the CRDs expose the standard conditions defined here, this library will handle them correctly.
+
+The `status` objects for built-in types are currently quite different. Not all built-in types expose conditions,
+and even among the types that does, the types of conditions vary widely. Long-term, we hope to add support for the
+standard conditions into the built-in types as well. This would remove the need for type-specific rules for determining
+status.
+
+## Statuses
+
+The library currently defines the following statuses for resource:
+* __InProgress__: The actual state of the resource has not yet reached the desired state as specified in the
+resource manifest, i.e. the resource reconcile has not yet completed.
+* __Failed__: The process of reconciling the actual state with the desired state has encountered and error
+or it has made insufficient progress.
+* __Current__: The actual state of the resource matches the desired state. The reconcile process is considered
+complete until there are changes to either the desired or the actual state.
+* __Terminating__: The resource is in the process of being deleted.
+* __Unknown__: This is for situations when the library are unable to determine the status of a resource.
```

Comment:
So will the new world _build_ on this (make it more complex) or _wrap_ it and make it simpler?

Reply:
So I think these statuses should be sufficient for finding out if the applied resources have fully reconciled. There might be other use cases where one needs other statuses. For example, if my use case was whether jobs had completed or were still running I would probably look at statuses like Running, Completed, Failed..

Reply:
This smells like a state machine.

To bookend "terminating" - does a resource undergoing creation have a detectable state? __InProgress__?

Should add a comment on this, and which status (status) is is entered upon creation.  Presumably can go right to __Failed__

Reply:
The library does not have any state machine. It just computes the status of each resource each time. It does not have any assumption about which transitions that would be allowed.

I have updated the docs with an additional comment about newly created resources.